### ISSUE TITLE
'Message waiting for you' on log-in

### ISSUE THREAD

**Issue by PatReynolds** [FreeUKGen/MyopicVicar#746]
To avoid email traffic, a 'message waiting for you' to be added to the log-in procedure - will enable us to contact all active volunteers

**Comment by Captainkirkdawson** [comment]
Requires the addition of an embedded collection of messages for a userid and a means of displaying them as well as a flag saying messages are available

**Comment by edickens** [comment]
Transcriber enhancement V1.6 or even Researcher Enhancement V1.7.
    
    But the message would not have been addressed to the UserID
    personally.  Even a general message would need to be flagged.  So
    just a simple system.  Any message that would have gone to that
    UserID needs the flag set against the UserID until the UserID has
    read the messages, or perhaps they can unset the flag themselves
    rather than having to check that they have read each individual
    message.  E
    
      On 15/10/2016 04:48, Kirk Dawson wrote:
    
    
      Requires the addition of an embedded collection of messages for
        a userid and a means of displaying them as well as a flag saying
        messages are available
      —
        You are receiving this because you are subscribed to this
        thread.
        Reply to this email directly, view
          it on GitHub, or mute
          the thread.
      
        
          
          
        
        
      
      {"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/FreeUKGen/MyopicVicar","title":"FreeUKGen/MyopicVicar","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/FreeUKGen/MyopicVicar"}},"updates":{"snippets":[{"icon":"PERSON","message":"@Captainkirkdawson in #746: Requires the addition of an embedded collection of messages for a userid and a means of displaying them as well as a flag saying messages are available"}],"action":{"name":"View Issue","url":"https://github.com/FreeUKGen/MyopicVicar/issues/746#issuecomment-253960558"}}}

**Comment by Captainkirkdawson** [comment]
As a result of this story
An embedded collection of message ids will be added for a userid
On login a message will be displayed that indicates there are messages waiting in the embedded collection for that userid with a link to the message display system .
The message display system will be able to display messages for a userid who has messages
When a message has been reviewed it will be removed from the userid embedded collection but not deleted from the messages collection itself.
When a message is sent to a userid the message id will be added to the embedded userid collection.

**Comment by edickens** [comment]
Looks great to me.  E
    
      On 15/10/2016 15:24, Kirk Dawson wrote:
    
    
      As a result of this story
        An embedded collection of message ids will be added for a userid
        On login a message will be displayed that indicates there are
        messages waiting in the embedded collection for that userid with
        a link to the message display system .
        The message display system will be able to display messages for
        a userid who has messages
        When a message has been reviewed it will be removed from the
        userid embedded collection but not deleted from the messages
        collection itself.
        When a message is sent to a userid the message id will be added
        to the embedded userid collection.
      —
        You are receiving this because you commented.
        Reply to this email directly, view
          it on GitHub, or mute
          the thread.
      
        
          
          
        
        
      
      {"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/FreeUKGen/MyopicVicar","title":"FreeUKGen/MyopicVicar","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/FreeUKGen/MyopicVicar"}},"updates":{"snippets":[{"icon":"PERSON","message":"@Captainkirkdawson in #746: As a result of this story\r\nAn embedded collection of message ids will be added for a userid\r\nOn login a message will be displayed that indicates there are messages waiting in the embedded collection for that userid with a link to the message display system .\r\nThe message display system will be able to display messages for a userid who has messages\r\nWhen a message has been reviewed it will be removed from the userid embedded collection but not deleted from the messages collection itself.\r\nWhen a message is sent to a userid the message id will be added to the embedded userid collection."}],"action":{"name":"View Issue","url":"https://github.com/FreeUKGen/MyopicVicar/issues/746#issuecomment-253987279"}}}

**Comment by israelriibeiro** [comment]
[branch issue_746](https://github.com/FreeUKGen/MyopicVicar/compare/issue_746)

Hi, @Captainkirkdawson . 

In order to resolve this, implemented this logic:

1- Every sent_message will have a checked set by default as false, which means that the user haven't seen the message yet.
2- In the view, when logged in, i added a message with link to index message. When the user login he'll see a message telling him how much messages does he have. I count this amount by calculating how much "unchecked" sent_message that user has. 
3- When the user sees the message, i update sent_message checked attribute to true.

The only problem that i didn't manage to resolve is this: if i try to send the same message twice, it  disappears from user's count. For exemplo, i sent to israel2. And, then, i do the same with israel3. It disappears from israel2.

![test_issue746](https://cloud.githubusercontent.com/assets/4265692/19984454/8d487a30-a1f5-11e6-8de8-47e4980277b1.gif)

**Comment by Captainkirkdawson** [comment]
Sorry Israel I am out of time. Passing you over to Ben/Doug

Kirk Dawson
5220 Riverside Drive
Fairmont Hot Springs, B.C.
V0B 1L1

**Comment by Sherlock21** [comment]
In the email below, is a snip of the screen image at logon.

On my email, this image is continually changing and moving up and down.

now: is that just a function of it being a little demo of what you are proposing will happen”
or is it a snag with copying the website page to the email?

if its a sign of things to come, then I am not happy to have my concentration  with the job I an doing, being interrupted with what i see here.

delving further into the workings: 
will the incoming mail sent message interrupt Any other screens? like if /i am in the middle of creating place names or relocation a file, then will i get this on they screen too?
Or only as I access the FR website on my browser.
Eric B

**Comment by israelriibeiro** [comment]
Hi, Eric. 

The gif it's a demo with i'm proposing. I already implemented this. 

This feature will not interrupt any action on the system. If you login to the system, you see that you got messages to check. That's it. I carefully followed what Pat said: 

> a 'message waiting for you' to be added to the log-in procedure[...]

Should i keep going through this path, or it needs some changes?

**Comment by Sherlock21** [comment]
Thanks, Israel

Lets wait and see when FR comes back then.

**Comment by richpomfret** [comment]
On hold while we decide on our approach for all messaging/communication related tasks.

**Comment by israelriibeiro** [comment]
I know this on hold, but i managed to discover which was the problem:

I created an has_many/belongs_to association between Useriddetail and Message, which means that, when that messages attribute( has_many :userid_messages - UseridDetails) is updated,  @message has its own userid_detail_id changed too [Update](https://github.com/FreeUKGen/MyopicVicar/blob/79d08204b5caaa969ba67f39689fe1c558e7020b/app/controllers/messages_controller.rb#L120) . Rails does that. Cause it's a has_many/belongs_to association, rails changes the id that belongs to another useriddetail object. That was my mistake.

Now i solved this problem this way: i created an array in useriddetails that keeps all messages id that belongs to him, if the messages is deleted, the array will be updated, cause when the user login, i check if the message was deleted, if it was a remove from this array. I made a gif for you to see.

Of course this is one approach of many, we could use has_many :through association, but this would require a huge change - another class to be more specific.

![test_msgs](https://cloud.githubusercontent.com/assets/4265692/20460270/25cbe068-aec5-11e6-9035-c5ffba725792.gif)

**Comment by PatReynolds** [comment]
Seems to be needed for the management of transcriber>coord>transcriber messages, whatever the outcome of the Gateway discussions

**Comment by Captainkirkdawson** [comment]
Me to review and push if OK

**Comment by Captainkirkdawson** [comment]
@israelriibeiro Pulled and checked. 

Message waiting message is there. However it is at the bottom of the page and hence very easily missed.  (I totally missed it first few times)

Appears I mislaid my earlier observations so putting in here.

1)Suggest the message waiting be moved to the top or we put into a page of its own and if there is a message waiting the manage_resources/new goes straight to it. The latter might be better especially in light of 2 below

2) The message link goes to the main editing page for messages it should not. It should go to a page that displays the messages and permits their deletion from the waiting list but not delete the message from the basic message system

3) It is not clear that the messages waiting is based on the receipt of a message that has been sent to a list of recipients. It seemed to me that it was based on the sender. 

4) the number of messages did not seem to be correct.

5) If zero one still goes to the display of messages.

6. Linking on the number is not a good idea. We need a bigger link.

**Comment by richpomfret** [comment]
Alison to test/review and suggest any improvements.

**Comment by AlOneill** [comment]
I can't *test* the message system as such -- I don't appear to have any messages to read!

As to how it should look, I suggest something along these lines. This would be the view on logging in.

![messages_on_actions_page](https://cloud.githubusercontent.com/assets/14852554/24909567/5babaf80-1ebc-11e7-8d78-fde4e0a5a45d.png)

What does the team think?

**Comment by Sherlock21** [comment]
It depends upon 
a) What kind of messages they are
b) Will this replace the messages being sent by email?
c) How will one reply to any of these messages that need a reply?
d) If the replay needs to go to several people, how will that be done?

**Comment by Captainkirkdawson** [comment]
@AlOneill Should have one now
@Sherlock21 a)they are only the messages from the Message system at this time b) no

**Comment by AlOneill** [comment]
Thanks @Captainkirkdawson . Now that I have seen how it works, I would drop the "to read" from the message.

I can also envisage that people will want to know a little more detail, especially if they tend to keep messages for future reference: how many new/unread messages as well as the total number. Perhaps that would be for a future enhancement, as more programming would be required.

**Comment by AlOneill** [comment]
This is what I see on logging in to test3 today:

![action_page___freereg](https://cloud.githubusercontent.com/assets/14852554/25488640/70189780-2b5f-11e7-94f1-34bc2dcc1ec3.png)

Are we really happy with the current wording? It is more street than business-like. And 'waiting' surely applies only if the messages have not yet been read. (They persist after being viewed, unless deleted by the user — as does the meta-message.)

Perhaps my mock-up (above) is a little over-elaborate, but I suggest the wording would be better as:
"You have 2 messages."
Then some space before the information that applies to all members. (This could be done by wrapping the 'You have n messages' in p tags.)

@Captainkirkdawson If you would like me make any edits, point me to the file and I'll do so.

**Comment by Captainkirkdawson** [comment]
I am not happy unless you are!

https://github.com/FreeUKGen/MyopicVicar/blob/2909763bcfc1d239b492ffa5c29274bf5e16460c/app/views/manage_resources/actions.html.erb

The userid messages view is

https://github.com/FreeUKGen/MyopicVicar/blob/master/app/views/messages/userid_messages.html.erb

K

Kirk Dawson
5220 Riverside Drive
Fairmont Hot Springs, B.C.
V0B 1L1

**Comment by edickens** [comment]
I have some messages waiting on Test3 so thought I'd have a look.  Some suggestions:-
- The confirmation message should be more visible.  Perhaps red, but not as pale as now..
![capture](https://cloud.githubusercontent.com/assets/8876165/25496986/f691d306-2b7a-11e7-91bd-87fe9f8043d8.JPG)
- The list of messages is in date order.  I suggest it should be most recent first, or the date column should be sortable.

But very useful.

**Comment by PatReynolds** [comment]
"Check them out" (or "check it/them out")?


In an ideal world, only new (i.e. unread) messages would be counted.  But
that is probably best as a new story for Version 2.

**Comment by PatReynolds** [comment]
Assigned to Anna for learning/information, not to work on this issue.

**Comment by AlOneill** [comment]
"Messages waiting" has been adjusted, including the text. The link colour will change when we adjust links globally.

Adjustments to listing in progress.

**Comment by Captainkirkdawson** [comment]
Adjustments made by Alison and deployed to production
